File: python/kswitch/service.py
"""B005.2 governed service API wrappers.

These SDK methods mirror the `kswitch_service` MCP surface. They are transport
wrappers only; policy, identity binding, audit persistence, and provider
dispatch remain server-side responsibilities.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .client import KSwitchAsyncClient, KSwitchClient


SERVICE_BASE = "/api/v1/b005/service"


def _compact(payload: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in payload.items() if value is not None}


class ServiceAPI:
    """Synchronous B005.2 governed service operations."""

    def __init__(self, client: KSwitchClient) -> None:
        self._c = client

    def fetch(self, *, url: str, purpose: str, task_id: str, max_bytes: int = 1048576) -> dict[str, Any]:
        return self._c._post(f"{SERVICE_BASE}/fetch", json={
            "url": url,
            "purpose": purpose,
            "task_id": task_id,
            "max_bytes": max_bytes,
        })

    def search(
        self,
        *,
        query: str,
        purpose: str,
        task_id: str,
        provider_id: str = "customer_search_default",
        max_results: int = 10,
    ) -> dict[str, Any]:
        return self._c._post(f"{SERVICE_BASE}/search", json={
            "query": query,
            "purpose": purpose,
            "task_id": task_id,
            "provider_id": provider_id,
            "max_results": max_results,
        })

    def policy_check(
        self,
        *,
        action: str,
        target: dict[str, Any],
        purpose: str,
        task_id: str,
        service_class: str | None = None,
    ) -> dict[str, Any]:
        return self._c._post(f"{SERVICE_BASE}/policy_check", json=_compact({
            "action": action,
            "target": target,
            "purpose": purpose,
            "task_id": task_id,
            "service_class": service_class,
        }))

    def get_policy(self) -> dict[str, Any]:
        return self._c._get(f"{SERVICE_BASE}/policy")

    def health(self) -> dict[str, Any]:
        return self._c._get(f"{SERVICE_BASE}/health")


class ServiceAsyncAPI:
    """Asynchronous B005.2 governed service operations."""

    def __init__(self, client: KSwitchAsyncClient) -> None:
        self._c = client

    async def fetch(self, *, url: str, purpose: str, task_id: str, max_bytes: int = 1048576) -> dict[str, Any]:
        return await self._c._post(f"{SERVICE_BASE}/fetch", json={
            "url": url,
            "purpose": purpose,
            "task_id": task_id,
            "max_bytes": max_bytes,
        })

    async def search(
        self,
        *,
        query: str,
        purpose: str,
        task_id: str,
        provider_id: str = "customer_search_default",
        max_results: int = 10,
    ) -> dict[str, Any]:
        return await self._c._post(f"{SERVICE_BASE}/search", json={
            "query": query,
            "purpose": purpose,
            "task_id": task_id,
            "provider_id": provider_id,
            "max_results": max_results,
        })

    async def policy_check(
        self,
        *,
        action: str,
        target: dict[str, Any],
        purpose: str,
        task_id: str,
        service_class: str | None = None,
    ) -> dict[str, Any]:
        return await self._c._post(f"{SERVICE_BASE}/policy_check", json=_compact({
            "action": action,
            "target": target,
            "purpose": purpose,
            "task_id": task_id,
            "service_class": service_class,
        }))

    async def get_policy(self) -> dict[str, Any]:
        return await self._c._get(f"{SERVICE_BASE}/policy")

    async def health(self) -> dict[str, Any]:
        return await self._c._get(f"{SERVICE_BASE}/health")

File: go/kswitch/service.go
package kswitch

import "context"

const serviceBasePath = "/api/v1/b005/service"

// ServiceService mirrors the B005.2 kswitch_service MCP tool surface.
//
// These methods are transport wrappers. Server-side policy, identity binding,
// audit persistence, and provider dispatch remain authoritative.
type ServiceService struct {
	client *Client
}

type ServiceFetchRequest struct {
	URL      string `json:"url"`
	Purpose  string `json:"purpose"`
	TaskID   string `json:"task_id"`
	MaxBytes int    `json:"max_bytes"`
}

type ServiceSearchRequest struct {
	Query      string `json:"query"`
	Purpose    string `json:"purpose"`
	TaskID     string `json:"task_id"`
	ProviderID string `json:"provider_id"`
	MaxResults int    `json:"max_results"`
}

type ServicePolicyCheckRequest struct {
	Action       string         `json:"action"`
	Target       map[string]any `json:"target"`
	Purpose      string         `json:"purpose"`
	TaskID       string         `json:"task_id"`
	ServiceClass string         `json:"service_class,omitempty"`
}

func (s *ServiceService) Fetch(ctx context.Context, req *ServiceFetchRequest) (map[string]any, error) {
	body := *req
	if body.MaxBytes == 0 {
		body.MaxBytes = 1048576
	}
	var out map[string]any
	err := s.client.do(ctx, "POST", serviceBasePath+"/fetch", &body, &out)
	return out, err
}

func (s *ServiceService) Search(ctx context.Context, req *ServiceSearchRequest) (map[string]any, error) {
	body := *req
	if body.ProviderID == "" {
		body.ProviderID = "customer_search_default"
	}
	if body.MaxResults == 0 {
		body.MaxResults = 10
	}
	var out map[string]any
	err := s.client.do(ctx, "POST", serviceBasePath+"/search", &body, &out)
	return out, err
}

func (s *ServiceService) PolicyCheck(ctx context.Context, req *ServicePolicyCheckRequest) (map[string]any, error) {
	var out map[string]any
	err := s.client.do(ctx, "POST", serviceBasePath+"/policy_check", req, &out)
	return out, err
}

func (s *ServiceService) GetPolicy(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.do(ctx, "GET", serviceBasePath+"/policy", nil, &out)
	return out, err
}

func (s *ServiceService) Health(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.do(ctx, "GET", serviceBasePath+"/health", nil, &out)
	return out, err
}
